## Introduction to Python for Machine Learning - Graded Quiz

## Question 1:
If you’re given two lists:

A = [1,2,3,4,5,6], B = [13, 21, 34]

The task is to create a list with the elements of A and B in a single dimension with output:

A_B =  [1,2,3,4,5,6,13, 21, 34]

In [38]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]

In [39]:
AB = A.extend(B)

In [40]:
print(A)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


In [41]:
A_B = A

In [42]:
print(A_B)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


Therefore, the right option is A.extend(B)

## Question 2:
Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [43]:
# import the pandas library
import pandas as pd

In [44]:
# importing the dataset into a dataframe
fuel_data = pd.read_csv('fuel_ferc1.csv')

In [46]:
# Finding the lowest average fuel cost per unit burned by grouping by the fuel type code
fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

From the above code, it can be concluded that 'gas' has the lowest average fuel cost per unit burned.

## Question 3:
What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [62]:
# Finding the skewness of the fuel quantity burned 
skewness = fuel_data.skew(axis = 0, skipna = True)
kurtosis = fuel_data.kurtosis(axis = 0, skipna = True)

print(skewness)

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64


In [63]:
print(kurtosis)

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64


From the above it can be concluded that the skweness and kurtosis of the fuel quantity burned are 15.85 and 651.37 respectively.

## Question 4:
Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places? 

(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)

In [64]:
# Finding the feature with missing value and the total number of missing value
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [76]:
#Percentage of the missing rows as a factor of the total number of rows 
percent_missing = 180 * 100 / fuel_data['fuel_unit'].count() 

In [77]:
print(percent_missing)

0.6134342091810653


It can be concluded that only the 'fuel_unit' feature has missing values with a total number of 180 missing values and 0.613 percent of the total number of rows are missing.

## Question 5:
What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [79]:
#Findind the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit)
fuel_data.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

From the above table it can be concluded that the standard deviation and 75th percentile of fuel_mmbtu_per_unit are 10.60 and 17.01 respectively.

## Question 6:
Which year has the highest average fuel cost per unit delivered?

In [82]:
#Finding the year with the highest average fuel cost per unit delivered

fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

From the above output it is seen clearly that the year 1997 had the highest average fuel cost per unit delivered with an average value of 11140.197239

## Question 7:
The feature with missing values falls under what category? What missing value imputation technique would you use?

The fuel_unit features falls under the numerical category and the MODE missing value imputation technique would work in filling the Null values.

In [83]:
#Filling the fuel_unit feature missing values with the modal class
fuel_data[['fuel_unit']] = fuel_data[['fuel_unit']].fillna(value='mcf')

In [84]:
# Confirming the missing values again
fuel_data.isnull().sum()

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

## Question 8:
Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [86]:
#Finding the correlation  values for all features
fuel_data.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

From the above, it can be concluded that 'fuel_qty_burned' and 'fuel_mmbtu_per_unit' has the second and third lowest correlation with Fuel cost per unit burned.

## Question 9:
For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [88]:
# Percentage change in the fuel cost per unit burned
fuel_data.groupby(['fuel_type_code_pudl', 'report_year'])['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl  report_year
coal                 1994            31.546467
                     1995            30.677442
                     1996            29.454299
                     1997           104.897309
                     1998            27.616234
                     1999            38.361101
                     2000           249.203055
                     2001            41.622623
                     2002            27.623988
                     2003            94.085441
                     2004            29.913239
                     2005            34.281426
                     2006            37.928697
                     2007            39.288397
                     2008            44.507292
                     2009           142.805197
                     2010           280.467547
                     2011            52.634721
                     2012            53.657789
                     2013            55.093312
                     2014  

From the above output, the fuel_cost_per_unit_burned for 1998 and 1994 are 27.616234 and 31.546467 respectively.

In [92]:
year_change =  27.616234 - 31.546467
percentage_change = year_change * 100 / 31.546467

print(percentage_change)

-12.458552014715313


From the above, percentage change in the fuel cost per unit burned for 1998 compared to 1994 is approximately -12%

## Question 10:
How do you create an identity matrix in python?

In [93]:
#creatng identity matrix in Python
import numpy as np 
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])